### PR TITLE
changes for paybc direct payment url

### PR DESCRIPTION
**Pull request by saravanpa-aot**
*Issue #:*
https://github.com/bcgov/entity/issues/4242

*Description of changes:*

- changes mainly for url parameters for paybc direct feature
_I added some TODO across the code where i am not sure if its right.Please have a look at them_

*Checklist:*
I confirm that below checklist items are addressed in this pull request; (check the boxes applicable)
- [ ] No lint errors
- [ ] No test case failures and proper coverage for all modules/classes
- [ ] Updated the deployment configs for new environment variable(s)
- [ ] Updated the postman collection in entity repository (https://github.com/bcgov/entity/tree/master/api-e2e/postman) for e2e tests

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the sbc-pay license (Apache 2.0).

### CODE REVIEW

File: pay-api/src/pay_api/services/direct_pay_service.py [hunk 1]
```diff
@@ -12,28 +12,89 @@
...
     def get_payment_system_url(self, invoice: Invoice, inv_ref: InvoiceReference, return_url: str):
         """Return the payment system url."""
-        # TODO return the proper url .Have to be addressed in the next task
-        return None
+        today = date.today().strftime(PAYBC_DATE_FORMAT)
+
+        url_params = {'trnDate': today,
+                      'pbcRefNumber': current_app.config.get('PAYBC_DIRECT_PAY_REF_NUMBER'),
+                      'glDate': today,
+                      'description': 'DIRECT SALE',
```
Comment (saravanpa-aot) [inline pay-api/src/pay_api/services/direct_pay_service.py:50]: I forgot what we decided to put as the description..do u remember it ? @sumesh-aot

File: pay-api/src/pay_api/services/direct_pay_service.py [hunk 2]
```diff
@@ -12,28 +12,89 @@
...
+                      'trnNumber': inv_ref.reference_number,
```
Comment (sumesh-aot) [inline pay-api/src/pay_api/services/direct_pay_service.py:51]: trnNumber should be the invoice_id, we need to make sure there is only 1 record in PayBC if it's direct pay and trnNumber is unique, and doesn't allow duplicate payment for same trnNumber. For PayBC invoiced CC flow, there can be multiple references as delete payment exists.

File: pay-api/src/pay_api/services/direct_pay_service.py [hunk 3]
```diff
@@ -12,28 +12,89 @@
...
+                      'trnAmount': invoice.total,
+                      'paymentMethod': PaymentMethod.CC.value,
+                      'redirectUri': return_url,
+                      'currency': 'CAD',
+                      'revenue': DirectPayService._create_revenue_string(invoice)}
+
+        url_params['hashValue'] = HashingService.encode(json.dumps(url_params))
```
Comment (sumesh-aot) [inline pay-api/src/pay_api/services/direct_pay_service.py:58]: Does it work? Basically we should be encoding the request params in URL after constructing the URL. If it works, good.

File: pay-api/src/pay_api/services/direct_pay_service.py [hunk 4]
```diff
@@ -12,28 +12,89 @@
...
+        paybc_url = current_app.config.get('PAYBC_DIRECT_PAY_PORTAL_URL')
+        return f'{paybc_url}?{encoded_query_params}'
+
+    @staticmethod
+    def _create_revenue_string(invoice) -> str:
+        payment_line_items = PaymentLineItemModel.find_by_invoice_ids([invoice.id])
+        index: int = 0
+        revenue_item = []
+        for payment_line_item in payment_line_items:
+            if payment_line_item.fee_schedule_id is not None:  # TODO is this logic needed???
```
Comment (sumesh-aot) [inline pay-api/src/pay_api/services/direct_pay_service.py:69]: Every line item should have a fee schedule id, so not really needed

File: pay-api/src/pay_api/services/direct_pay_service.py [hunk 5]
```diff
@@ -12,28 +12,89 @@
...
+                distribution_code = DistributionCodeModel.find_by_active_for_fee_schedule(
```
Comment (sumesh-aot) [inline pay-api/src/pay_api/services/direct_pay_service.py:70]: You can get the distribution from fee_distribution_id in payment line item

File: pay-api/src/pay_api/services/direct_pay_service.py [hunk 6]
```diff
@@ -53,7 +114,8 @@ def create_invoice(self, payment_account: PaymentAccount, line_items: [PaymentLi
         current_app.logger.debug('<create_invoice_direct_pay')
 
         invoice = {
-            'invoice_number': f'{invoice.id}'
+            'invoice_number': f'{invoice.id}',
+            'reference_number': secrets.randbits(32)  # TODO is this necessary here..or whats the correct logic
```
Comment (sumesh-aot) [inline pay-api/src/pay_api/services/direct_pay_service.py:118]: Not really needed, check my comment on top.

File: pay-api/tests/unit/services/test_hashing_service.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
+# Copyright © 2019 Province of British Columbia
```
Comment (sumesh-aot) [inline pay-api/tests/unit/services/test_hashing_service.py:1]: Please add a test_direct_pay_service as it involves more logic.

File: pay-api/src/pay_api/services/direct_pay_service.py (hunk 1 above)
Comment (sumesh-aot) [reply pay-api/src/pay_api/services/direct_pay_service.py:50]: It's not displayed anywhere. May be make it as `Direct Sale`.

Comment (saravanpa-aot) [reply pay-api/src/pay_api/services/direct_pay_service.py]: this is done

File: pay-api/src/pay_api/services/direct_pay_service.py (hunk 2 above)
Comment (saravanpa-aot) [reply pay-api/src/pay_api/services/direct_pay_service.py]: I changed the txnNumber .Now it is 'trnNumber': invoice.id,

File: pay-api/src/pay_api/services/direct_pay_service.py (hunk 3 above)
Comment (saravanpa-aot) [reply pay-api/src/pay_api/services/direct_pay_service.py]: Fixed it.

File: pay-api/src/pay_api/services/direct_pay_service.py (hunk 4 above)
Comment (saravanpa-aot) [reply pay-api/src/pay_api/services/direct_pay_service.py]: removed it

File: pay-api/src/pay_api/services/direct_pay_service.py (hunk 5 above)
Comment (saravanpa-aot) [reply pay-api/src/pay_api/services/direct_pay_service.py]: changed

File: pay-api/tests/unit/services/test_hashing_service.py (hunk 7 above)
Comment (saravanpa-aot) [reply pay-api/tests/unit/services/test_hashing_service.py:1]: working on it.

**Pull request merged**